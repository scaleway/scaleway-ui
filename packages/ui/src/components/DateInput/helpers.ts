import { CURRENT_MONTH, CURRENT_YEAR } from './constants'

// First day of the month for a given year
export const getMonthFirstDay = (
  month = CURRENT_MONTH,
  year = CURRENT_YEAR,
) => {
  const firstDay = new Date(year, month - 1, 1).getDay()

  // Change so that a week starts on monday
  return firstDay === 0 ? 6 : firstDay - 1
}

export const addZero = (value: number) => `${value}`.padStart(2, '0')

export const getPreviousMonth = (month: number, year: number) => {
  if (month === 1) {
    return [12, year - 1]
  }

  return [month - 1, year]
}

export const getNextMonth = (month: number, year: number) => {
  if (month === 12) {
    return [1, year + 1]
  }

  return [month + 1, year]
}

// Checks if two date values are of the same month and year
export const isSameMonth = (date: Date, basedate = new Date()) =>
  basedate.getMonth() === date.getMonth() &&
  basedate.getFullYear() === date.getFullYear()

// (bool) Checks if two date values are the same day
export const isSameDay = (date: Date, basedate = new Date()) =>
  basedate.getDate() === date.getDate() &&
  basedate.getMonth() + 1 === date.getMonth() + 1 &&
  basedate.getFullYear() === date.getFullYear()

// Default format if none is provided
const getDateISO = (showMonthYearPicker: boolean, date?: Date) => {
  if (date) {
    if (showMonthYearPicker) {
      return [addZero(date.getMonth() + 1), date.getFullYear()].join('/')
    }

    return [
      addZero(date.getMonth() + 1),
      addZero(date.getDate()),
      date.getFullYear(),
    ].join('/')
  }

  return ''
}

export const formatValue = (
  computedValue: Date | null,
  computedRange: {
    start: Date | null
    end: Date | null
  } | null,
  showMonthYearPicker: boolean,
  selectsRange: boolean,
  format?: (value?: Date) => string | undefined,
) => {
  if (selectsRange && computedRange) {
    return format
      ? `${format(computedRange.start ?? undefined) ? `${format(computedRange.start ?? undefined)} - ` : ''}${format(computedRange.end ?? undefined) ?? ''}`
      : `${getDateISO(showMonthYearPicker, computedRange.start ?? undefined)}${computedRange.start ? ' - ' : ''}${getDateISO(showMonthYearPicker, computedRange.end ?? undefined)}`
  }

  if (computedValue && format) {
    return format(computedValue)
  }
  if (computedValue) {
    return getDateISO(showMonthYearPicker, computedValue)
  }

  return undefined
}
