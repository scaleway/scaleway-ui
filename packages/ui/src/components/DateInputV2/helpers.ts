export const POPUP_WIDTH = '264px'

// The current year
export const CURRENT_YEAR = new Date().getFullYear()

// The current month starting from 1 - 12
// 1 => January, 12 => December
export const CURRENT_MONTH = new Date().getMonth() + 1

// The current day 1-28/29/30/31

// Weeks displayed on calendar
export const CALENDAR_WEEKS = 6

// Number days in a month for a given year
export const getMonthDays = (month = CURRENT_MONTH, year = CURRENT_YEAR) => {
  const months30 = [4, 6, 9, 11] // 30 days in april, june, september and november
  const leapYear = year % 4 === 0

  if (month === 2) {
    return leapYear ? 29 : 28 // Days in february
  }

  return months30.includes(month) ? 30 : 31
}

// First day of the month for a given year from 1 (monday) - 7 (saturday)
export const getMonthFirstDay = (month = CURRENT_MONTH, year = CURRENT_YEAR) =>
  new Date(`${year}-${month}-01`).getDay() + 1

// Adds a zero for 1-digit months/days
export const addZero = (value: number) => `${value}`.padStart(2, '0')

export const getPreviousMonth = (month: number, year: number) => {
  if (month === 1) {
    return [12, year - 1]
  }

  return [month - 1, year]
}

export const getNextMonth = (month: number, year: number) => {
  if (month === 12) {
    return [1, year + 1]
  }

  return [month + 1, year]
}

// Checks if two date values are of the same month and year
export const isSameMonth = (date: Date, basedate = new Date()) =>
  basedate.getMonth() + 1 === date.getMonth() + 1 &&
  basedate.getFullYear() === date.getFullYear()

// (bool) Checks if two date values are the same day
export const isSameDay = (date: Date, basedate = new Date()) =>
  basedate.getDate() === date.getDate() &&
  basedate.getMonth() + 1 === date.getMonth() + 1 &&
  basedate.getFullYear() === date.getFullYear()
