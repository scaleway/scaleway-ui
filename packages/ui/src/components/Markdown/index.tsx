import { css } from '@emotion/react'
import styled from '@emotion/styled'
import type { ReactNode } from 'react'
import ReactMarkDown from 'react-markdown'
import Link from '../Link'
import Text from '../Text'

const StyledLink = styled(Link)`
  font-size: inherit;
`

const StyledText = styled(Text)`
  margin-top: 0;
`

const Container = styled.div`
  display: inline-flex;
  background-color: ${({ theme }) => theme.colors.neutral.backgroundStrong};
  padding: ${({ theme }) => theme.space['1']};
`

const MarDownHeading = ({
  children,
  ...props
}: {
  level: number
  children: ReactNode
}) => {
  if (props.level === 1) {
    return (
      <StyledText variant="heading" as="h2">
        {children}
      </StyledText>
    )
  }
  const { heading: Heading } = ReactMarkDown.renderers

  // eslint-disable-next-line react/jsx-props-no-spreading
  return <Heading {...props}>{children}</Heading>
}

const MarkDownCode = ({ children }: { children: ReactNode }) => (
  <Container>
    <Text as="code" variant="code">
      {children}
    </Text>
  </Container>
)
const MdText = ({ children }: { children: ReactNode }) => (
  <span>{children}</span>
)
const MarkDownParagraph = ({ children }: { children: ReactNode }) => (
  <p>{children}</p>
)

const MarkDownLink = ({
  children,
  href,
}: {
  href: string
  children: ReactNode
}) => {
  if (!href) {
    return null
  }

  return (
    <StyledLink variant="info" href={href}>
      {children}
    </StyledLink>
  )
}

const StyledContainer = styled('div', {
  shouldForwardProp: prop => !['inline'].includes(prop),
})<{ inline?: boolean }>`
  ${({ inline }) =>
    inline &&
    css`
      > p {
        display: inline;
      }
    `}
`

type MarkDownContainerProps = Pick<MarkDownProps, 'inline' | 'className'>

const rootRenderer =
  ({ inline, className }: MarkDownContainerProps) =>
  ({ children }: { children: ReactNode[] }) =>
    (
      <StyledContainer inline={inline} className={className}>
        {children}
      </StyledContainer>
    )

type MarkDownProps = {
  source: string
  linkTarget?: string
  escapeHtml?: boolean
  inline?: boolean
  className?: string
}

/**
 * @Deprecated in favor of Text component
 */
const Markdown = ({
  source,
  linkTarget,
  escapeHtml = true,
  inline,
  className,
}: MarkDownProps) => (
  <ReactMarkDown
    source={source}
    renderers={{
      heading: MarDownHeading,
      inlineCode: MarkDownCode,
      link: MarkDownLink,
      paragraph: MarkDownParagraph,
      root: rootRenderer({ inline, className }),
      text: MdText,
    }}
    linkTarget={linkTarget}
    escapeHtml={escapeHtml}
  />
)

export default Markdown
