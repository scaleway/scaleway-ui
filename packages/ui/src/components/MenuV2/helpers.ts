import { Children, type ReactNode, isValidElement } from 'react'

/**
 * Search inside a children (React Element) recursively until a result is found
 */
export const searchChildren = (
  children: ReactNode,
  searchString: string,
): ReactNode[] => {
  const matches: ReactNode[] = []
  const searchRegex = new RegExp(searchString, 'i')

  Children.forEach(children, child => {
    if (typeof child === 'string' && child.match(searchRegex)) {
      matches.push(child)
    } else if (isValidElement(child)) {
      const childProps = child.props as { children: ReactNode }

      // Recursively search the children of this element
      const childMatches = searchChildren(childProps.children, searchString)

      if (childMatches.length > 0) {
        // If any matches are found within this child's children, push the entire child element
        matches.push(child)
      }
    }
  })

  return matches
}
