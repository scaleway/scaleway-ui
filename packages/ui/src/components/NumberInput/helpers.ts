export const bounded = (value: number, min: number, max: number) =>
  Math.max(min, Math.min(value, max))

export const roundStep = (value: number, step: number, direction: number) =>
  direction === -1
    ? Math.floor(value / step) * step
    : Math.ceil(value / step) * step

export const getMinusRoundedValue = (currentValue: number, step: number) =>
  currentValue % step === 0
    ? roundStep(currentValue - step, step, -1)
    : roundStep(currentValue, step, -1)

export const getPlusRoundedValue = (currentValue: number, step: number) =>
  currentValue % step === 0
    ? roundStep(currentValue + step, step, 1)
    : roundStep(currentValue, step, 1)
