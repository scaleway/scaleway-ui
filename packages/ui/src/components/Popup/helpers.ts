import type { RefObject } from 'react'

export type PopupPlacement = 'top' | 'right' | 'bottom' | 'left' | 'auto'
export const ARROW_WIDTH = 8 // in px
const SPACE = 4 // in px
const TOTAL_USED_SPACE = 0 // in px
export const DEFAULT_POSITIONS = {
  arrowLeft: -999,
  arrowTop: -999,
  arrowTransform: 'translate(-50%, -50)',
  placement: 'top',
  rotate: 135,
  popupInitialPosition: 'translate3d(-999px, -999px, 0)',
  popupPosition: 'translate3d(-999px, -999px, 0)',
}

type ComputePlacementTypes = {
  childrenStructuredRef: DOMRect
  popupStructuredRef: DOMRect
  popupPortalTarget: HTMLElement
  offsetParentRect: DOMRect
  offsetParent: Element
}

/**
 * This function will find the best placement in a window for popup based on children position and popup size
 */
const computePlacement = ({
  childrenStructuredRef,
  popupStructuredRef,
  offsetParentRect,
  offsetParent,
  popupPortalTarget,
}: ComputePlacementTypes) => {
  const {
    top: childrenTop,
    left: childrenLeft,
    right: childrenRight,
  } = childrenStructuredRef

  const {
    top: parentTop,
    left: parentLeft,
    right: parentRight,
  } = offsetParentRect

  const isPopupPortalTargetBody =
    popupPortalTarget === document.body || offsetParent === document.body

  const overloadedChildrenLeft = isPopupPortalTargetBody
    ? childrenLeft
    : childrenLeft - parentLeft
  const overloadedChildrenTop = isPopupPortalTargetBody
    ? childrenTop
    : childrenTop - parentTop
  const overloadedChildrenRight = isPopupPortalTargetBody
    ? childrenRight
    : childrenRight - parentRight

  const { width: popupWidth, height: popupHeight } = popupStructuredRef

  if (overloadedChildrenTop - popupHeight - TOTAL_USED_SPACE < 0) {
    return 'bottom'
  }

  if (overloadedChildrenLeft - popupWidth - TOTAL_USED_SPACE < 0) {
    return 'right'
  }

  if (
    overloadedChildrenRight + popupWidth + TOTAL_USED_SPACE >
    window.innerWidth
  ) {
    return 'left'
  }

  return 'top'
}

/**
 * This function will check if the offset parent is usable for popup positioning
 * If not it will loop and search for a compatible parent until document.body is reached
 */
const findOffsetParent = (element: RefObject<HTMLDivElement>) => {
  const offsetParent = element?.current?.offsetParent

  // We need to check if offsetParent is a table cell or a table because they are not suitable for positioning
  // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetParent
  if (offsetParent && !['TH', 'TD', 'TABLE'].includes(offsetParent.tagName)) {
    return offsetParent
  }

  let currentElement = element?.current

  while (currentElement && currentElement.tagName !== 'BODY') {
    const { position } = window.getComputedStyle(currentElement)

    // Check if the current element is a potential offset parent
    if (position !== 'static') {
      return currentElement
    }

    currentElement = currentElement.parentElement as HTMLDivElement
  }

  // If no suitable offset parent is found, return the body element
  return document.body
}

type ComputePositionsTypes = {
  placement: PopupPlacement
  childrenRef: RefObject<HTMLDivElement>
  popupRef: RefObject<HTMLDivElement>
  popupPortalTarget: HTMLElement
}

/**
 * This function will compute the positions of popup and arrow based on children position and popup size
 */
export const computePositions = ({
  placement,
  childrenRef,
  popupRef,
  popupPortalTarget,
}: ComputePositionsTypes) => {
  const childrenRect = (
    childrenRef.current as HTMLDivElement
  ).getBoundingClientRect()
  const offsetParent = findOffsetParent(childrenRef)
  const offsetParentRect = offsetParent?.getBoundingClientRect() ?? {
    top: 0,
    left: 0,
    right: 0,
  }
  const popupStructuredRef = (
    popupRef.current as HTMLDivElement
  ).getBoundingClientRect()

  const placementBasedOnWindowSize =
    placement === 'auto'
      ? computePlacement({
          childrenStructuredRef: childrenRect,
          popupStructuredRef,
          offsetParentRect,
          popupPortalTarget,
          offsetParent,
        })
      : placement

  const {
    top: childrenTop,
    left: childrenLeft,
    right: childrenRight,
    width: childrenWidth,
    height: childrenHeight,
  } = childrenRect

  const {
    top: parentTop,
    left: parentLeft,
    right: parentRight,
  } = offsetParentRect

  const { width: popupWidth, height: popupHeight } = popupStructuredRef

  // offSetParent is the closest positioned ancestor. If the element is not positioned, the nearest table cell or root element is used.
  const isPopupPortalTargetBody =
    popupPortalTarget === document.body || offsetParent === document.body

  // It will get how much scroll is done on the page to compute the position of the popup
  const scrollTopValue = isPopupPortalTargetBody
    ? document.documentElement.scrollTop
    : offsetParent.scrollTop

  // We need to compute the position of the popup based on the parent element in the case the popup is not in the body
  const overloadedChildrenLeft = isPopupPortalTargetBody
    ? childrenLeft
    : childrenLeft - parentLeft
  const overloadedChildrenTop = isPopupPortalTargetBody
    ? childrenTop
    : childrenTop - parentTop
  const overloadedChildrenRight = isPopupPortalTargetBody
    ? childrenRight
    : childrenRight + childrenWidth + ARROW_WIDTH + SPACE - parentRight / 2

  switch (placementBasedOnWindowSize) {
    case 'bottom': {
      const positionX =
        overloadedChildrenLeft + childrenWidth / 2 - popupWidth / 2
      const positionY =
        overloadedChildrenTop +
        scrollTopValue +
        childrenHeight +
        ARROW_WIDTH +
        SPACE

      return {
        arrowLeft: popupWidth / 2,
        arrowTop: -ARROW_WIDTH - 5,
        arrowTransform: '',
        placement: 'bottom',
        rotate: 180,
        popupInitialPosition: `translate3d(${positionX}px, ${
          positionY - TOTAL_USED_SPACE
        }px, 0)`,
        popupPosition: `translate3d(${positionX}px, ${positionY}px, 0)`,
      }
    }
    case 'left': {
      const positionX =
        overloadedChildrenLeft - popupWidth - ARROW_WIDTH - SPACE * 2
      const positionY =
        overloadedChildrenTop +
        scrollTopValue -
        popupHeight / 2 +
        childrenHeight / 2

      return {
        arrowLeft: popupWidth + ARROW_WIDTH + 5,
        arrowTop: popupHeight / 2,
        arrowTransform: 'translate(-50%, -50%)',
        placement: 'left',
        rotate: -90,
        popupInitialPosition: `translate3d(${
          positionX + TOTAL_USED_SPACE
        }px, ${positionY}px, 0)`,
        popupPosition: `translate3d(${positionX}px, ${positionY}px, 0)`,
      }
    }
    case 'right': {
      const positionX = overloadedChildrenRight + ARROW_WIDTH + SPACE * 2
      const positionY =
        overloadedChildrenTop +
        scrollTopValue -
        popupHeight / 2 +
        childrenHeight / 2

      return {
        arrowLeft: -ARROW_WIDTH - 5,
        arrowTop: popupHeight / 2,
        arrowTransform: 'translate(50%, -50%)',
        placement: 'right',
        rotate: 90,
        popupInitialPosition: `translate3d(${
          positionX - TOTAL_USED_SPACE
        }px, ${positionY}px, 0)`,
        popupPosition: `translate3d(${positionX}px, ${positionY}px, 0)`,
      }
    }
    default: {
      // top placement is default value
      const positionX =
        overloadedChildrenLeft + childrenWidth / 2 - popupWidth / 2
      const positionY =
        overloadedChildrenTop +
        scrollTopValue -
        popupHeight -
        ARROW_WIDTH -
        SPACE

      return {
        arrowLeft: popupWidth / 2,
        arrowTop: popupHeight - 1,
        arrowTransform: '',
        placement: 'top',
        rotate: 0,
        popupInitialPosition: `translate3d(${positionX}px, ${
          positionY + TOTAL_USED_SPACE
        }px, 0)`,
        popupPosition: `translate3d(${positionX}px, ${positionY}px, 0)`,
      }
    }
  }
}
