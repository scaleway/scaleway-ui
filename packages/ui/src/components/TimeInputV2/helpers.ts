const format = (value: string, timeFormat: 12 | 24, isHour?: boolean) => {
  if (value === '') return ''
  if (timeFormat === 12 && isHour) {
    const formattedValue = Number.parseInt(value, 10) % 12
    const finalValue = formattedValue === 0 ? 12 : formattedValue

    return finalValue.toString().padStart(2, '0')
  }

  return value.padStart(2, '0')
}

const computePeriod = (hours?: string, period?: 'am' | 'pm') => {
  if (period) return period

  if (hours && Number.parseInt(hours, 10) >= 12) return 'pm'

  return 'am'
}

export const isNumber = (key: string) => /^\d+$/.test(key)
export const isAOrP = (key: string) => ['a', 'A', 'p', 'P'].includes(key)

const isValidNewHour = (old: string, key: string, timeFormat: number) => {
  const newValue = Number.parseInt(old + key, 10)

  return (
    (timeFormat === 24 && newValue < 24) || (timeFormat === 12 && newValue < 13)
  )
}

export const isCompleteHour = (timeFormat: number, value?: string) => {
  const computedValue = Number.parseInt(value ?? '0', 10)

  if (timeFormat === 12 && computedValue > 1) return true
  if (timeFormat === 24 && computedValue > 2) return true

  return false
}

export const canConcat = (
  oldValue: string | undefined,
  type: 'h' | 'm' | 's',
  key: string,
  timeFormat: 24 | 12,
) =>
  oldValue &&
  ((type === 'h' && oldValue && isValidNewHour(oldValue, key, timeFormat)) ||
    (type !== 'h' && Number.parseInt(oldValue, 10) < 6))

export const formatValue = (
  value: {
    h: string
    m: string
    s: string
    period?: string
  },
  timeFormat: 24 | 12,
) => ({
  h: format(value.h, timeFormat, true),
  m: format(value.m, timeFormat),
  s: format(value.s, timeFormat),
  period: computePeriod(
    value.h,
    value.period === 'am' || value.period === 'pm' ? value.period : undefined,
  ),
})
