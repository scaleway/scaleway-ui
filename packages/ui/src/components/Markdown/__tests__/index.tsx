import Markdown from '..'
import { shouldMatchEmotionSnapshot } from '../../../helpers/jestHelpers'

export const simple = `
  # I am a Markdown H1

  ## I am a Markdown H2

  I am a simple markdown with **strong** _italic_ content

  You can find the complete reference [here](https://commonmark.org/help/)

  Invalid [link]()

  I can also render code

  \`const example = 1 // Inline code\`

  \`\`\`
    const example = 2
  \`\`\`
`

export const withMultipleParagraphs = `
  I am a paragraph.

  And a second one
`

export const withHtml = `
  Hello<br />World
`

describe('MarkDown', () => {
  test(`render simple string`, () =>
    shouldMatchEmotionSnapshot(<Markdown source={simple} />))

  test(`render simple string with target blank`, () =>
    shouldMatchEmotionSnapshot(
      <Markdown source={simple} linkTarget="_blank" />,
    ))

  test(`render paragraph inline`, () =>
    shouldMatchEmotionSnapshot(
      <Markdown source={withMultipleParagraphs} inline />,
    ))

  test(`render html without escaing it`, () =>
    shouldMatchEmotionSnapshot(
      <Markdown source={withHtml} escapeHtml={false} />,
    ))
})
