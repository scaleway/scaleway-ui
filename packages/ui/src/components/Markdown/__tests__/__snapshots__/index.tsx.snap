// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkDown render html without escaping it 1`] = `
<DocumentFragment>
  <div
    class="cache-18ai6pl-StyledContainer e1kf57p60"
  >
    <p>
      <span>
        Hello
      </span>
      <br />
      <span>
        World
      </span>
    </p>
  </div>
</DocumentFragment>
`;

exports[`MarkDown render paragraph inline 1`] = `
<DocumentFragment>
  .cache-rus7tn-StyledContainer-StyledContainer>p {
  display: inline;
}

<div
    class="cache-rus7tn-StyledContainer-StyledContainer e1kf57p60"
  >
    <p>
      <span>
        I am a paragraph.
      </span>
    </p>
    <p>
      <span>
        And a second one
      </span>
    </p>
  </div>
</DocumentFragment>
`;

exports[`MarkDown render simple string 1`] = `
<DocumentFragment>
  .cache-8pvx2m-StyledText-StyledText {
  color: #4a4f62;
  font-size: 25px;
  font-family: Asap;
  font-weight: 400;
  letter-spacing: 0;
  line-height: 32px;
  text-transform: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin-top: 0;
}

.cache-1ns6cco-StyledLink-StyledLink {
  background-color: transparent;
  border: none;
  padding: 0;
  color: #1a4dbf;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  text-decoration-thickness: 1px;
  text-underline-offset: 2px;
  text-decoration-color: transparent;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: text-decoration-color 250ms ease-out;
  transition: text-decoration-color 250ms ease-out;
  gap: 8px;
  position: relative;
  cursor: pointer;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  font-size: 16px;
  font-family: Asap;
  font-weight: 500;
  letter-spacing: 0;
  line-height: 24px;
  paragraph-spacing: 0;
  text-case: none;
  font-size: inherit;
}

.cache-1ns6cco-StyledLink-StyledLink .e4bq1r52 {
  -webkit-transition: -webkit-transform 250ms ease-out;
  transition: transform 250ms ease-out;
}

.cache-1ns6cco-StyledLink-StyledLink>* {
  pointer-events: none;
}

.cache-1ns6cco-StyledLink-StyledLink:hover,
.cache-1ns6cco-StyledLink-StyledLink:focus {
  outline: none;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  text-decoration-thickness: 1px;
  color: #1a4dbf;
  text-decoration-color: #1a4dbf;
}

.cache-1ns6cco-StyledLink-StyledLink:hover .e4bq1r52,
.cache-1ns6cco-StyledLink-StyledLink:focus .e4bq1r52 {
  -webkit-transform: translate(-4px, 0);
  -moz-transform: translate(-4px, 0);
  -ms-transform: translate(-4px, 0);
  transform: translate(-4px, 0);
}

.cache-1ns6cco-StyledLink-StyledLink:hover::after,
.cache-1ns6cco-StyledLink-StyledLink:focus::after {
  background-color: #1a4dbf;
}

.cache-1ns6cco-StyledLink-StyledLink:active {
  text-decoration-thickness: 2px;
}

.cache-46pynu-Container {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  background-color: #f6f6f8;
  padding: 8px;
}

.cache-1v48bjl-StyledText {
  color: #4a4f62;
  font-size: 14px;
  font-family: JetBrains;
  font-weight: 400;
  letter-spacing: 0;
  line-height: 20px;
  text-transform: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
    class="cache-18ai6pl-StyledContainer e1kf57p60"
  >
    <h2
      class="e1kf57p62 cache-8pvx2m-StyledText-StyledText e1tg3t120"
    >
      <span>
        I am a Markdown H1
      </span>
    </h2>
    <h2>
      <span>
        I am a Markdown H2
      </span>
    </h2>
    <p>
      <span>
        I am a simple markdown with 
      </span>
      <strong>
        <span>
          strong
        </span>
      </strong>
      <span>
         
      </span>
      <em>
        <span>
          italic
        </span>
      </em>
      <span>
         content
      </span>
    </p>
    <p>
      <span>
        You can find the complete reference 
      </span>
      <a
        class="e1kf57p63 cache-1ns6cco-StyledLink-StyledLink e4bq1r50"
        href="https://commonmark.org/help/"
      >
        <span>
          here
        </span>
      </a>
    </p>
    <p>
      <span>
        Invalid 
      </span>
    </p>
    <p>
      <span>
        I can also render code
      </span>
    </p>
    <p />
    <div
      class="cache-46pynu-Container e1kf57p61"
    >
      <code
        class="cache-1v48bjl-StyledText e1tg3t120"
      >
        const example = 1 // Inline code
      </code>
    </div>
    <p />
    <pre>
      <code>
          const example = 2
      </code>
    </pre>
  </div>
</DocumentFragment>
`;

exports[`MarkDown render simple string with target blank 1`] = `
<DocumentFragment>
  .cache-8pvx2m-StyledText-StyledText {
  color: #4a4f62;
  font-size: 25px;
  font-family: Asap;
  font-weight: 400;
  letter-spacing: 0;
  line-height: 32px;
  text-transform: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin-top: 0;
}

.cache-1ns6cco-StyledLink-StyledLink {
  background-color: transparent;
  border: none;
  padding: 0;
  color: #1a4dbf;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  text-decoration-thickness: 1px;
  text-underline-offset: 2px;
  text-decoration-color: transparent;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: text-decoration-color 250ms ease-out;
  transition: text-decoration-color 250ms ease-out;
  gap: 8px;
  position: relative;
  cursor: pointer;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  font-size: 16px;
  font-family: Asap;
  font-weight: 500;
  letter-spacing: 0;
  line-height: 24px;
  paragraph-spacing: 0;
  text-case: none;
  font-size: inherit;
}

.cache-1ns6cco-StyledLink-StyledLink .e4bq1r52 {
  -webkit-transition: -webkit-transform 250ms ease-out;
  transition: transform 250ms ease-out;
}

.cache-1ns6cco-StyledLink-StyledLink>* {
  pointer-events: none;
}

.cache-1ns6cco-StyledLink-StyledLink:hover,
.cache-1ns6cco-StyledLink-StyledLink:focus {
  outline: none;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  text-decoration-thickness: 1px;
  color: #1a4dbf;
  text-decoration-color: #1a4dbf;
}

.cache-1ns6cco-StyledLink-StyledLink:hover .e4bq1r52,
.cache-1ns6cco-StyledLink-StyledLink:focus .e4bq1r52 {
  -webkit-transform: translate(-4px, 0);
  -moz-transform: translate(-4px, 0);
  -ms-transform: translate(-4px, 0);
  transform: translate(-4px, 0);
}

.cache-1ns6cco-StyledLink-StyledLink:hover::after,
.cache-1ns6cco-StyledLink-StyledLink:focus::after {
  background-color: #1a4dbf;
}

.cache-1ns6cco-StyledLink-StyledLink:active {
  text-decoration-thickness: 2px;
}

.cache-46pynu-Container {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  background-color: #f6f6f8;
  padding: 8px;
}

.cache-1v48bjl-StyledText {
  color: #4a4f62;
  font-size: 14px;
  font-family: JetBrains;
  font-weight: 400;
  letter-spacing: 0;
  line-height: 20px;
  text-transform: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
    class="cache-18ai6pl-StyledContainer e1kf57p60"
  >
    <h2
      class="e1kf57p62 cache-8pvx2m-StyledText-StyledText e1tg3t120"
    >
      <span>
        I am a Markdown H1
      </span>
    </h2>
    <h2>
      <span>
        I am a Markdown H2
      </span>
    </h2>
    <p>
      <span>
        I am a simple markdown with 
      </span>
      <strong>
        <span>
          strong
        </span>
      </strong>
      <span>
         
      </span>
      <em>
        <span>
          italic
        </span>
      </em>
      <span>
         content
      </span>
    </p>
    <p>
      <span>
        You can find the complete reference 
      </span>
      <a
        class="e1kf57p63 cache-1ns6cco-StyledLink-StyledLink e4bq1r50"
        href="https://commonmark.org/help/"
      >
        <span>
          here
        </span>
      </a>
    </p>
    <p>
      <span>
        Invalid 
      </span>
    </p>
    <p>
      <span>
        I can also render code
      </span>
    </p>
    <p />
    <div
      class="cache-46pynu-Container e1kf57p61"
    >
      <code
        class="cache-1v48bjl-StyledText e1tg3t120"
      >
        const example = 1 // Inline code
      </code>
    </div>
    <p />
    <pre>
      <code>
          const example = 2
      </code>
    </pre>
  </div>
</DocumentFragment>
`;
